### ISSUE TITLE
dask_cudf xgboost memory leak

### ISSUE THREAD

**Issue by pseudotensor** [dmlc/xgboost#6280]
Same setup as here: #6232 https://github.com/dmlc/xgboost/issues/6232

```
import pandas as pd
def fun():
    from dask.distributed import Client, wait
    from dask_cuda import LocalCUDACluster

    with LocalCUDACluster() as cluster:
        with Client(cluster) as client:

            import xgboost as xgb
            import dask_cudf

            target = "default payment next month"
            Xpd = pd.read_csv("creditcard.csv")
            Xpd = Xpd[['AGE', target]]
            Xpd.to_csv("creditcard_1.csv")
            X = dask_cudf.read_csv("creditcard_1.csv")
            y = X[target]
            X = X.drop(target, axis=1)

            kwargs_fit = {}
            kwargs_cudf_fit = kwargs_fit.copy()

            valid_X = dask_cudf.read_csv("creditcard_1.csv")
            valid_y = valid_X[target]
            valid_X = valid_X.drop(target, axis=1)
            kwargs_cudf_fit['eval_set'] = [(valid_X, valid_y)]

            params = {}  # copy.deepcopy(self.model.get_params())
            params['tree_method'] = 'gpu_hist'
            params['eval_metric'] = 'auc'

            i = 0
            while True:
                dask_model = xgb.dask.DaskXGBClassifier(**params)
                dask_model.fit(X, y, verbose=True)
                preds = dask_model.predict(X).compute().ravel()
                print("here %d: %s" % (i, preds[0]))
                del dask_model
                i += 1


if __name__ == '__main__':
    fun()

```

Running this uses an ever-growing amount of memory.  It's about 0.5MB per cycle on GPU even for a single GPU.

i.e. I'm staring at the dask-worker, which here started out at 400MB and is after about 40 cycles at 420MB:

![image](https://user-images.githubusercontent.com/2249614/96938446-440c7000-147f-11eb-8789-532e41dd074e.png)

**Comment by trivialfis** [comment]
I can't reproduce it.  With memory logging inside xgboost, the peak memory usage is always on 1MB.

**Comment by trivialfis** [comment]
@pseudotensor Could you please try the latest nsight system?  It can track memory usage.

**Comment by trivialfis** [comment]
I sort of reproduce it.  But it's more a Python thing instead of issues in libraries.  For documents on memory management with dask, see https://distributed.dask.org/en/latest/memory.html .  `client.restart` is actually a valid way to purge unused states.  Also for simple training with XGBoost inside loops, Python `gc` is quite useful from my testing.  Without calling `gc.collect` the memory usage keeps growing.  But forcing gc can keep the memory usage stable.

``` python
import xgboost as xgb
from xgboost import dask as dxgb
from distributed import Client, wait
from dask_cuda import LocalCUDACluster
from dask import dataframe as dd
import dask_cudf
import cudf
import numpy as np
from time import time
import os
import gc

fname = "HIGGS.csv"
colnames = ["label"] + ["feature-%02d" % i for i in range(1, 29)]


def dmain():
    with LocalCUDACluster() as cluster:
        print("Dashboard link:", cluster.dashboard_link)
        with Client(cluster) as client:
            dask_df = dask_cudf.read_csv(fname, header=None, names=colnames)
            y = dask_df["label"]
            X = dask_df[dask_df.columns.difference(["label"])]

            for i in range(0, 100):
                print("i:", i)
                dtrain = dxgb.DaskDMatrix(client, X, y)
                start = time()
                output = dxgb.train(
                    client, {"tree_method": "gpu_hist"}, dtrain, num_boost_round=10
                )
                end = time()
                print("Train::Duration", end - start)

                start = time()
                predictions = dxgb.predict(client, output, dtrain)
                predictions = client.persist(predictions)
                wait(predictions)
                end = time()

                del predictions
                del dtrain
                gc.collect()

                print("Predict::Duration", end - start)
